system_metrics_collector/include/moving_average_statistics/moving_average.hpp

```diff
@@ -0,0 +1,189 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MOVING_AVERAGE_STATISTICS__MOVING_AVERAGE_HPP_
+#define MOVING_AVERAGE_STATISTICS__MOVING_AVERAGE_HPP_
+
+
+#include <cmath>
+
+#include <algorithm>
+#include <mutex>
+#include <type_traits>
+
+#include "moving_average_statistics/types.hpp"
+
+
+/**
+ *  A class for calculating moving average statistics. This operates in constant memory and constant time. Note:
+ *  reset() must be called manually in order to start a new measurement window.
+ *
+ *  The statistics calculated are average, maximum, minimum, and standard deviation (population).
+ *  All are calculated online without storing the observation data. Specifically, the average is a running sum
+ *  and the variance is obtained by Welford's online algorithm
+ *  (reference: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford%27s_online_algorithm)
+ *  for standard deviation.
+ *
+ *  When statistics are not available, e.g. no observations have been made, NaNs are returned.
+ *
+ *  @tparam T the type to be observed. Note: this must conform to std::is_arithmetic.
```

Comment:
Why can't we use always double here? That would simplify the code a lot and get rid of the template.

Reply:
I agree. We had the case before to try and extend (with multiple stats collectors), but I don't see how that use case applies here anymore.